File: src/main/java/com/mmust/evaluators/impl/ShuntingYard.java
package com.mmust.evaluators.impl;

import com.mmust.token.Associativity;
import com.mmust.token.Operator;
import com.mmust.token.Token;
import com.mmust.token.TokenType;

import java.util.Deque;
import java.util.LinkedList;
import java.util.Queue;

class ShuntingYard {

	private Deque<Token> operatorStack = new LinkedList<>();
	private Queue<Token> outputQueue = new LinkedList<>();

	private Iterable<Token> infixExpression;

	ShuntingYard(Iterable<Token> infixExpression) {
		this.infixExpression = infixExpression;
	}

	Iterable<Token> doSorting() {
		for (Token token : infixExpression) {
			TokenType type = token.getType();
			if (type == TokenType.NUMBER) {
				outputQueue.add(token);
			} else if (type == TokenType.FUNCTION) {
				operatorStack.add(token);
			} else if (type == TokenType.OPERATOR) {
				Operator op = (Operator) token;
				while (
						isFunctionOnTop() || isHigherPriorityOpOnTop(op) || isLeftAssociativeAndEqualPriorityOpOnTop(op)
								&& !isLeftParenthesisOnTop()
				) {
					outputQueue.add(operatorStack.pop());
				}
				operatorStack.push(token);
			} else if (type == TokenType.LEFT_PARENTHESIS) {
				operatorStack.push(token);
			} else if (type == TokenType.RIGHT_PARENTHESIS) {
				while (!isLeftParenthesisOnTop()) {
					outputQueue.add(operatorStack.pop());
				}
				if (isLeftParenthesisOnTop()) {
					operatorStack.pop();
				}
			}
		}

		outputQueue.addAll(operatorStack);

		return outputQueue;
	}

	private boolean isLeftAssociativeAndEqualPriorityOpOnTop(Operator op) {
		Token top = operatorStack.peek();
		if (top != null && top.getType() == TokenType.OPERATOR) {
			Operator topOp = (Operator) top;
			return topOp.getAssociativity() == Associativity.LEFT && op.getPrecedence() == topOp.getPrecedence();
		}
		return false;
	}

	private boolean isHigherPriorityOpOnTop(Operator op) {
		if (!operatorStack.isEmpty()) {
			Token top = operatorStack.peek();
			if (top.getType() == TokenType.OPERATOR) {
				Operator topOp = (Operator) top;
				return topOp.getPrecedence() > op.getPrecedence();
			}
		}
		return false;
	}

	private boolean isFunctionOnTop() {
		return !operatorStack.isEmpty() && operatorStack.peek().getType() == TokenType.FUNCTION;
	}

	private boolean isLeftParenthesisOnTop() {
		return !operatorStack.isEmpty() && operatorStack.peek().getType() == TokenType.LEFT_PARENTHESIS;
	}
}

File: src/main/kotlin/com/mmust/evaluators/impl/ShuntingYard.kt
package com.mmust.evaluators.impl

import com.mmust.token.*
import com.mmust.token.Function
import java.util.*

internal class ShuntingYard(private val infixExpression: Iterable<Token>) {

    private val operatorStack = LinkedList<Token>()
    private val outputQueue = LinkedList<Token>()

    private val isFunctionOnTop: Boolean
        get() = !operatorStack.isEmpty() && operatorStack.peek() is Function

    private val isLeftParenthesisOnTop: Boolean
        get() = !operatorStack.isEmpty() && operatorStack.peek() is Parenthesis && operatorStack.peek() == Parenthesis.LEFT

    fun doSorting(): Iterable<Token> {
        for (token in infixExpression) {
            when (token) {
                is NumToken -> outputQueue.add(token)
                is Function -> operatorStack.push(token)
                is BinaryOperator -> {
                    val op = token as BinaryOperator
                    while (isFunctionOnTop || isHigherPriorityOpOnTop(op) || isLeftAssociativeAndEqualPriorityOpOnTop(op) && !isLeftParenthesisOnTop) {
                        outputQueue.add(operatorStack.pop())
                    }
                    operatorStack.push(token)
                }
                is Parenthesis -> {
                    if (token == Parenthesis.LEFT) {
                        operatorStack.push(token)
                    } else {
                        while (!isLeftParenthesisOnTop) {
                            outputQueue.add(operatorStack.pop())
                        }
                        if (isLeftParenthesisOnTop) {
                            operatorStack.pop()
                        }
                    }
                }
            }
        }

        outputQueue.addAll(operatorStack)

        return outputQueue
    }

    private fun isLeftAssociativeAndEqualPriorityOpOnTop(op: BinaryOperator): Boolean {
        val top = operatorStack.peek()
        if (top != null && top is BinaryOperator) {
            val topOp = top as BinaryOperator
            return topOp.associativity == Associativity.LEFT && op.precedence == topOp.precedence
        }
        return false
    }

    private fun isHigherPriorityOpOnTop(op: BinaryOperator): Boolean {
        if (!operatorStack.isEmpty()) {
            val top = operatorStack.peek()
            if (top is BinaryOperator) {
                val topOp = top as BinaryOperator
                return topOp.precedence > op.precedence
            }
        }
        return false
    }
}
